### PR TITLE
Fix nested-session auth leak in fcc-claude launcher; size auto-compact window to local models

### PR DESCRIPTION
**Pull request by jankeydadondc-byte**
## Summary

`fcc-claude`'s launcher env-builder (`build_claude_launcher_env` in
`cli/launchers/claude.py`) only strips `ANTHROPIC_*`-prefixed vars before
handing the child environment to `claude.exe`. If `fcc-claude` is invoked
from *inside* an already-running Claude Code session — for example, an
agent driving its own shell/Bash tool, which is a real and increasingly
common way people use fcc for local-model agent loops — the child process
inherits the parent session's identity vars: `CLAUDECODE`,
`CLAUDE_CODE_CHILD_SESSION`, `CLAUDE_CODE_SDK_HAS_HOST_AUTH_REFRESH`,
`CLAUDE_CODE_OAUTH_SCOPES`, `CLAUDE_AGENT_SDK_VERSION`, `CLAUDE_EFFORT`,
`AI_AGENT`, `BAGGAGE`.

With those present, the nested `claude.exe` decides it's a host-managed
child session and tries to authenticate via the parent's real Anthropic
OAuth credentials instead of respecting `ANTHROPIC_AUTH_TOKEN` — so every
request 401s against the local proxy, regardless of how `ANTHROPIC_AUTH_TOKEN`
/`ANTHROPIC_API_KEY` are set. This is invisible from the proxy side: the
token is configured correctly, the proxy is healthy, and a plain `curl` with
the same token succeeds — only the nested `claude.exe`/`fcc-claude` fails.

## Repro (before this fix)

```bash
# 1. fcc-server running normally, ANTHROPIC_AUTH_TOKEN=freecc in ~/.fcc/.env
fcc-server &

# 2. A direct curl with the configured token succeeds fine:
curl -s -o /dev/null -w "%{http_code}\n" -X POST http://localhost:8082/v1/messages \
  -H "x-api-key: freecc" -H "content-type: application/json" \
  -H "anthropic-version: 2023-06-01" \
  -d '{"model":"claude-3-sonnet","max_tokens":10,"messages":[{"role":"user","content":"hi"}]}'
# -> 200

# 3. But simulate invoking fcc-claude from inside a live Claude Code session
#    (i.e. with the parent session's own identity vars already in the
#    environment, as they would be for any nested Bash-tool invocation):
CLAUDECODE=1 CLAUDE_CODE_CHILD_SESSION=1 CLAUDE_CODE_SDK_HAS_HOST_AUTH_REFRESH=1 \
CLAUDE_CODE_OAUTH_SCOPES="user:inference user:profile" \
  fcc-claude -p "reply with just the word OK"
# -> Failed to authenticate. API Error: 401 {"detail":"Invalid API key"}
# fcc-server's own log confirms the request never even reaches with a
# matching token: LMSTUDIO/native `x-api-key`/`Authorization` header check
# never matches "freecc" because claude.exe substitutes its own bearer
# token instead of honoring ANTHROPIC_AUTH_TOKEN.
```

After this fix, step 3 succeeds identically to step 2 — no manual `env -u`
workaround needed.

## Fix

- `build_claude_launcher_env` now strips every inherited `CLAUDE_CODE_*`-
  prefixed var plus the fixed set above, then re-sets only the two
  `CLAUDE_CODE_*` vars the launcher itself intentionally configures
  (`CLAUDE_CODE_ENABLE_GATEWAY_MODEL_DISCOVERY`,
  `CLAUDE_CODE_AUTO_COMPACT_WINDOW`). `fcc-claude`'s contract is a fresh,
  standalone session pointed at the local proxy — any inherited
  session/harness var undermines that contract.
- Also adds `resolve_auto_compact_window()`: when the configured provider
  is `lmstudio`, it queries `GET <lm_studio_base_url>/api/v0/models` for
  the currently loaded model's `loaded_context_length` and sizes
  `CLAUDE_CODE_AUTO_COMPACT_WINDOW` to it (85% proportional reserve, or a
  flat 12k-token reserve at small contexts, whichever is smaller), instead
  of the static `190000` default. A locally-loaded model's real context
  ceiling is usually far below 190k tokens, so without this Claude Code
  never auto-compacts before hitting the local ceiling, and LM Studio
  silently truncates the stream mid-response instead of erroring cleanly.
  Falls back to the static default for every other provider, or on any
  lookup failure (fails open — never blocks a launch).

## Testing

- `uv run ruff format` / `uv run ruff check --fix` / `uv run ty check`:
  all clean on both changed files.
- Added 4 new unit tests in `tests/cli/test_entrypoints.py` covering the
  session-var stripping, the lmstudio-branch window sizing (mocked
  `httpx.get`), the non-lmstudio fallback, and the fail-open behavior on a
  lookup error. Existing `build_claude_launcher_env` tests are
  unmodified and still pass under the new signature (the new
  `auto_compact_window` parameter is optional, defaulting to the old
  static behavior).
- Note: I hit a pre-existing, environment-specific crash running the full
  suite locally on Windows (`OPENSSL_Uplink ... no OPENSSL_Applink`,
  inside `api.admin_urls`'s import chain) — confirmed unrelated to this
  change (reproduces identically on a clean, unmodified checkout of
  `main`). Verified the new/changed logic by direct execution of the
  functions outside pytest instead. Flagging in case it's a known issue
  with this Python/OpenSSL combination on Windows — happy to help debug
  further if useful.
- Version bumped `2.4.4` → `2.4.5` per `AGENTS.md`'s versioning rule (bug
  fix, no breaking change); `uv.lock` regenerated accordingly (note: a
  handful of unrelated `sys_platform` marker lines also changed in
  `uv.lock` — that's `uv lock` resolving with a newer local `uv`
  (0.11.26) than last touched the committed lockfile, not anything from
  this change; happy to hand-revert those lines if you'd rather keep the
  lockfile diff minimal).

## Scope

Single, focused fix — first of a small series I'm planning to send from
the same debugging session (this one's the least controversial / most
universally applicable of the set).

## Versioning note

The semver bump is intentionally omitted from this PR. `main` merges several
version-bumping PRs a day, so any version this branch pins goes stale (and
re-conflicts the PR) within hours — this branch has been rebased over version
conflicts alone several times already. The code change itself never conflicts.
Happy to push a bump at merge time if you ping me, or feel free to bump on
merge — whichever you prefer.

<h3>Greptile Summary</h3>

This PR updates the `fcc-claude` launcher to start nested Claude Code sessions with a clean proxy-focused environment. The main changes are:

- Strip inherited `ANTHROPIC_*`, `CLAUDE_CODE_*`, and Claude session identity variables before launching `claude.exe`.
- Re-set only the launcher-owned Claude Code environment variables needed for gateway model discovery and auto-compaction.
- Size `CLAUDE_CODE_AUTO_COMPACT_WINDOW` from loaded LM Studio model context lengths when any configured chat model uses LM Studio.
- Add tests for session variable stripping, LM Studio lookup behavior, model override handling, smallest-context sizing, small-context clamping, and lookup failure fallback.

<h3>Confidence Score: 4/5</h3>

The changed launcher flow is safe to merge after the repository versioning requirement is handled.

The launcher behavior is covered by focused tests and the previously reported auto-compact edge cases are addressed. Confidence is limited because `AGENTS.md` and `CLAUDE.md` require a semver bump and regenerated `uv.lock` for production changes under `src/free_claude_code/cli/`.

`pyproject.toml` and `uv.lock` need attention for the required version bump. The changed files do not need special attention.

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- Ran the Pytest suite for the launcher-auth-auto-compact target and confirmed 6 tests passed in 3.20 seconds with exit code 0.
- Validated Ruff format for the launcher-auth-auto-compact target, noting that 2 files were already formatted and exit code was 0.
- Verified Ruff lint checks for the launcher-auth-auto-compact target passed with all checks successful and exit code 0.
- Validated Ty type-check for the launcher-auth-auto-compact target, confirming all checks passed with exit code 0.

<a href="https://app.greptile.com/trex/runs/13880349/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/free_claude_code/cli/launchers/claude.py | Adds inherited Claude session env stripping and LM Studio-aware auto-compact window sizing; no blocking implementation issues found. |
| tests/cli/test_entrypoints.py | Adds tests for session variable stripping, LM Studio sizing, override routing, smallest-context selection, small-context clamping, and fail-open lookup behavior. |

<h3>Sequence Diagram</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
participant User
participant Launcher as fcc-claude launcher
participant Env as Child env builder
participant LM as LM Studio /api/v0/models
participant Claude as claude.exe
User->>Launcher: run fcc-claude
Launcher->>Launcher: load settings and proxy URL
alt any configured chat model uses lmstudio
    Launcher->>LM: GET /api/v0/models
    LM-->>Launcher: loaded_context_length values
    Launcher->>Launcher: choose smallest loaded context and compute window
else no lmstudio or lookup fails
    Launcher->>Launcher: use static compact window
end
Launcher->>Env: build_claude_launcher_env(base_env, token, window)
Env->>Env: "drop ANTHROPIC_*, CLAUDE_CODE_*, and session identity vars"
Env-->>Launcher: "proxy auth/base URL and launcher-owned CLAUDE_CODE_* vars"
Launcher->>Claude: start with sanitized environment
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
participant User
participant Launcher as fcc-claude launcher
participant Env as Child env builder
participant LM as LM Studio /api/v0/models
participant Claude as claude.exe
User->>Launcher: run fcc-claude
Launcher->>Launcher: load settings and proxy URL
alt any configured chat model uses lmstudio
    Launcher->>LM: GET /api/v0/models
    LM-->>Launcher: loaded_context_length values
    Launcher->>Launcher: choose smallest loaded context and compute window
else no lmstudio or lookup fails
    Launcher->>Launcher: use static compact window
end
Launcher->>Env: build_claude_launcher_env(base_env, token, window)
Env->>Env: "drop ANTHROPIC_*, CLAUDE_CODE_*, and session identity vars"
Env-->>Launcher: "proxy auth/base URL and launcher-owned CLAUDE_CODE_* vars"
Launcher->>Claude: start with sanitized environment
```

</a>

<sub>Reviews (12): Last reviewed commit: ["Fix nested-session auth leak in fcc-clau..."](https://github.com/alishahryar1/free-claude-code/commit/ae48d3e61af8ec48eae0682ae77ef3faf58ac5ad) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41863454)</sub>

### CODE REVIEW

File: src/free_claude_code/cli/launchers/claude.py [hunk 1]
```diff
@@ -67,21 +71,81 @@ def build_claude_launcher_command(
...
+        loaded = [
+            model.get("loaded_context_length")
+            for model in response.json().get("data", [])
+            if model.get("state") == "loaded"
+            and isinstance(model.get("loaded_context_length"), int)
+        ]
+        if not loaded:
+            return CLAUDE_CODE_AUTO_COMPACT_WINDOW
+        context_length = max(loaded)
+        window = max(8192, min(int(context_length * 0.85), context_length - 12000))
```
[the remark this answers is not in the document]
Comment (jankeydadondc-byte) [reply src/free_claude_code/cli/launchers/claude.py:140]: Good catch — fixed by clamping the final window to `context_length` (commit aa87798), so a small loaded context (e.g. 4096) now correctly returns 4096 instead of the 8192 floor overriding it. Added a regression test (`test_resolve_auto_compact_window_clamps_to_a_small_loaded_context`) covering exactly this case; verified the two existing cases (40960->28960, 130048->110540) are unchanged.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: cli/launchers/claude.py [hunk 2]
```diff
@@ -67,21 +71,87 @@ def build_claude_launcher_command(
...
+    """Match Claude Code's auto-compact window to the loaded local model.
+
+    The static default (190000) is far above a locally-loaded context like
+    45k/130k, so Claude Code never compacts before the local ceiling and LM
+    Studio silently truncates the stream. When LM Studio is the provider, read
+    the actually-loaded context length and reserve headroom for one large
+    tool-result turn plus generation.
+    """
+    if parse_provider_type(settings.model) != "lmstudio":
+        return CLAUDE_CODE_AUTO_COMPACT_WINDOW
```
[the remark this answers is not in the document]
Comment (jankeydadondc-byte) [reply cli/launchers/claude.py]: Fixed (commit 36f9b26): resolve_auto_compact_window now checks every configured slot via configured_chat_model_refs() and sizes the window if *any* of MODEL/MODEL_OPUS/MODEL_SONNET/MODEL_HAIKU routes to lmstudio — so an LM Studio override under a non-LM-Studio default gets a truthful window instead of the static 190000. Added a regression test for exactly that case. Sizing to the LM Studio ceiling is the safe choice since the window is a single global env var: missing compaction truncates silently, while applying it to a large-context cloud tier only compacts slightly early.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: cli/launchers/claude.py [hunk 3]
```diff
@@ -67,21 +71,98 @@ def build_claude_launcher_command(
...
+        loaded = [
+            model.get("loaded_context_length")
+            for model in response.json().get("data", [])
+            if model.get("state") == "loaded"
+            and isinstance(model.get("loaded_context_length"), int)
+        ]
+        if not loaded:
+            return CLAUDE_CODE_AUTO_COMPACT_WINDOW
+        context_length = max(loaded)
+        window = max(8192, min(int(context_length * 0.85), context_length - 12000))
```
[the remark this answers is not in the document]
Comment (jankeydadondc-byte) [reply cli/launchers/claude.py]: Fixed (commit b00d59d): switched max(loaded) to min(loaded). Since the compact window is a single global env var applied to every request, with contexts like [4096, 130048] it must fit the smallest — sizing to the largest left the 4096 model unprotected. min() protects the smallest-context model; applying that same window to a larger model only compacts slightly early, which is harmless. Common single-model case is unchanged (min == max). Added a multi-loaded-model regression test.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by jankeydadondc-byte** [comment]
@Alishahryar1 — rebased cleanly onto the new `src/free_claude_code/` layout (#1029); green on all required checks (Ban suppressions, ruff-format, ruff-check, ty, pytest) and currently **mergeable**. Scope is two related fixes — the nested-session auth leak in `fcc-claude` and sizing the auto-compact window to the loaded LM Studio context — plus their unit tests.

Same versioning note as #978: I've left the `pyproject.toml` bump out on purpose, because pinning a version just re-conflicts the PR every time `main` bumps. Happy to add the bump right before you merge, or you can bump on merge — your call.

Would appreciate a review when you get a chance. Thanks for maintaining this!

_🤖 Posted via automation on behalf of the PR author._

**Comment by jankeydadondc-byte** [comment]
Withdrawing this PR per our own review-cadence policy — no maintainer bandwidth issue implied. We'll keep maintaining the fix in our local patch set; happy to reopen or rebase if there's interest later. Thanks for maintaining the project!
